adala/runtimes/_litellm.py

```diff
@@ -603,109 +600,194 @@ def get_cost_estimate(
             )
 
 
-class LiteLLMVisionRuntime(LiteLLMChatRuntime):
+class MessageChunkType(Enum):
+    TEXT = "text"
+    IMAGE_URL = "image_url"
+
+
+def split_message_into_chunks(
+    input_template: str, input_field_types: Dict[str, MessageChunkType], **input_fields
+) -> List[MessageChunk]:
+    """Split a template string with field types into a list of message chunks.
+
+    Takes a template string with placeholders and splits it into chunks based on the field types,
+    preserving the text between placeholders.
+
+    Args:
+        input_template (str): Template string with placeholders, e.g. '{a} is a {b} is an {a}'
+        input_field_types (Dict[str, MessageChunkType]): Dict mapping field names to their types
+        **input_fields: Field values to substitute into template
+
+    Returns:
+        List[Dict[str, str]]: List of message chunks with appropriate type and content.
+            Text chunks have format: {'type': 'text', 'text': str}
+            Image chunks have format: {'type': 'image_url', 'image_url': {'url': str}}
+
+    Example:
+        >>> split_message_into_chunks(
+        ...     '{a} is a {b} is an {a}',
+        ...     {'a': MessageChunkType.TEXT, 'b': MessageChunkType.IMAGE_URL},
+        ...     a='the letter a',
+        ...     b='http://example.com/b.jpg'
+        ... )
+        [
+            {'type': 'text', 'text': 'the letter a is a '},
+            {'type': 'image_url', 'image_url': {'url': 'http://example.com/b.jpg'}},
+            {'type': 'text', 'text': ' is an the letter a'}
+        ]
+    """
+    # Parse template to get field positions and surrounding text
+    parsed = parse_template(input_template)
+    chunks: List[MessageChunk] = []
+
+    current_chunk: Optional[MessageChunk] = None
+
+    def add_to_current_chunk(chunk):
+        nonlocal current_chunk
```

Comment:
I'm guessing you've tried already, but would definitely like to avoid all this nonlocal business... I'd think we could accomplish the `push_current_chunk` behavior by using a generator function and iterating through that? And then for `add_to_current_chunk` we could accumulate into a var within the generator function, and empty it after `yield`ing?